### PR TITLE
Add Operator to Read the Newest Version of an Object

### PR DESCRIPTION
**Pull request by cancan101**
Based on Alias Operator

Can be used to solve https://github.com/pimcore/pimcore/issues/5322

CC @fashxp 

Guided by: https://github.com/orgs/pimcore/discussions/17056
We pass `false`, per https://github.com/pimcore/pimcore/pull/17059 as we already have object if published and can avoid a disk access.

Will be further improved with: https://github.com/pimcore/pimcore/issues/17728 and https://github.com/pimcore/pimcore/issues/17729 (and more so if [Delegate](https://pimcore.com/docs/platform/Pimcore/Tools_and_Features/Versioning/#delegate) is used)

The first row has a draft version:
image
Grid Options using the new operator:
image

### CODE REVIEW

**Comment by robertSt7** [comment]
@cancan101 Could you please use `const` and `let` instead of `var`? Thanks

**Comment by cancan101** [comment]
done

File: src/DataObject/GridColumnConfig/Operator/VersionGetter.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                }
+            }
+
+            $result->isArrayType = $isArrayType;
+            if ($isArrayType) {
+                $result->value = $valueArray;
+            } else {
+                $result->value = $valueArray[0] ?? null;
+            }
+        }
```
Comment (jdreesen) [inline src/DataObject/GridColumnConfig/Operator/VersionGetter.php]: There's no need to put this into an `else` block.

```suggestion
        }

        $c = $children[0];

        $valueArray = [];

        $latestVersion = $element->getLatestVersion(null, false);
        if($latestVersion){
            $element = $latestVersion->loadData();
        }

        $childResult = $c->getLabeledValue($element);
        $isArrayType = $childResult->isArrayType ?? null;
        $childValues = $childResult->value;
        if ($childValues && !$isArrayType) {
            $childValues = [$childValues];
        }

        if ($childValues) {
            /** @var string $childValue */
            foreach ($childValues as $childValue) {
                $valueArray[] = $childValue;
            }
        }

        $result->isArrayType = $isArrayType;
        if ($isArrayType) {
            $result->value = $valueArray;
        } else {
            $result->value = $valueArray[0] ?? null;
        }
```

**Comment by cancan101** [comment]
I can't see what sonarcloud is reporting without logging in to it.

File: public/js/pimcore/object/gridcolumn/operator/VersionGetter.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return this.window;
+    },
+
+    commitData: function (params) {
+        this.node.data.configAttributes.label = this.textfield.getValue();
+        this.node.set('text', this.textfield.getValue());
+        this.node.set('isOperator', true);
+
+        this.window.close();
+        if (params && params.callback) {
```
Comment (robertSt7) [inline public/js/pimcore/object/gridcolumn/operator/VersionGetter.js]:
```suggestion
        if (params?.callback) {
```
fix for SonarCloud

**Comment by robertSt7** [comment]
@cancan101 We need to add the file path of the VersionGetter [here](https://github.com/pimcore/admin-ui-classic-bundle/blob/b3d14d13bc9dba26a53354a6deb8d138cc1d355b/templates/admin/index/index.html.twig#L490) and I got the following error when I try to add it to the event folder grid. Could you please take care of this? Thanks 
```Status: 500 | Internal Server Error
URL: /admin/object/grid-proxy?classId=EV&folderId=1119
Method: POST
Message: Operator "VersionGetter" is not supported
Trace: 
in /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/DataObject/GridColumnConfig/Service.php:85
#0 /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/DataObject/GridColumnConfig/Service.php(72): Pimcore\Bundle\AdminBundle\DataObject\GridColumnConfig\Service->buildOperator('VersionGetter', Object(stdClass), Array)
#1 /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/DataObject/GridColumnConfig/Service.php(50): Pimcore\Bundle\AdminBundle\DataObject\GridColumnConfig\Service->doBuildConfig(Array, Array, Array)
#2 /var/www/html/dev/pimcore/pimcore/models/DataObject/Service.php(520): Pimcore\Bundle\AdminBundle\DataObject\GridColumnConfig\Service->buildOutputDataConfig(Array, Array)
#3 /var/www/html/dev/pimcore/pimcore/models/DataObject/Service.php(534): Pimcore\Model\DataObject\Service::getConfigForHelperDefinition(Array, '#674dd40e7bb67', Array)
#4 /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/Service/GridData/DataObject.php(81): Pimcore\Model\DataObject\Service::calculateCellValue(Object(Pimcore\Model\DataObject\Event), Array, '#674dd40e7bb67', Array)
#5 /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/Controller/Admin/DataObject/DataObjectActionsTrait.php(132): Pimcore\Bundle\AdminBundle\Service\GridData\DataObject::getData(Object(Pimcore\Model\DataObject\Event), Array, 'en', Array)
#6 /var/www/html/dev/pimcore/admin-ui-classic-bundle/src/Controller/Admin/DataObject/DataObjectController.php(1717): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->gridProxy(Array, 'object', Object(Symfony\Component\HttpFoundation\Request), Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher), Object(Pimcore\Bundle\AdminBundle\Helper\GridHelperService), Object(Pimcore\Localization\LocaleService))
#7 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(181): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->gridProxyAction(Object(Symfony\Component\HttpFoundation\Request), Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher), Object(Pimcore\Bundle\AdminBundle\Helper\GridHelperService), Object(Pimcore\Localization\LocaleService), Object(Pimcore\Bundle\AdminBundle\Security\CsrfProtectionHandler))
#8 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(76): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#9 /var/www/html/vendor/symfony/http-kernel/Kernel.php(197): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#10 /var/www/html/vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php(35): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#11 /var/www/html/vendor/autoload_runtime.php(29): Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner->run()
#12 /var/www/html/public/index.php(20): require_once('/var/www/html/v...')
#13 {main}```

**Comment by cancan101** [comment]
What is the "event folder grid"?

**Comment by robertSt7** [comment]
> What is the "event folder grid"?

@cancan101 Thanks for adding the stuff. I mean the folder in the demo, but I got the error also with every other folder.
image

**Comment by cancan101** [comment]
Fixed:
image

**Comment by robertSt7** [comment]
@cancan101 Thanks for the improvement

**Pull request merged**